packages/labs/compiler/src/test/source_map_test.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {SourceMapConsumer} from 'source-map';
+
+import {test} from 'uvu';
+// eslint-disable-next-line import/extensions
+import * as assert from 'uvu/assert';
+import {readFile} from 'fs/promises';
+import * as url from 'url';
+import * as path from 'path';
+
+const __dirname = url.fileURLToPath(new URL('.', import.meta.url));
+const TEST_FILES_DIR = path.resolve(
+  __dirname,
+  '../test_files/source_map_tests'
+);
+
+/**
+ * This helper asserts that a given position in the compiled file maps to a
+ * specific position in the authored file.
+ *
+ * To write these tests, use the editor to inspect line and column numbers in
+ * the source_map_tests directory.
+ *
+ * @param smc SourceMapConsumer
+ * @param generatedPosition The [line, column] for the generatedPosition.
+ * @param authoredPosition The [line, column] for the authoredPosition. May be
+ * null if there is no mapping.
+ */
+const assertSourceMapLocationMapping = ({
+  smc,
+  generatedPosition,
+  authoredPosition,
+  msg,
+}: {
+  smc: SourceMapConsumer;
+  generatedPosition: [number, number];
+  authoredPosition: [number | null, number | null];
+  msg: string;
+}) => {
+  const result = smc.originalPositionFor({
+    line: generatedPosition[0],
+    column: generatedPosition[1],
+  });
+
+  assert.equal([result.line, result.column], authoredPosition, msg);
+};
+
+const setupTest = async (filename: string): Promise<SourceMapConsumer> => {
+  const resolvedSourcemapPath = path.resolve(TEST_FILES_DIR, filename);
+  const map = await readFile(resolvedSourcemapPath, {encoding: 'utf-8'});
+  const smc = await new SourceMapConsumer(map);
+  return smc;
+};
+
+test('basic.ts', async () => {
+  const smc = await setupTest('basic.js.map');
+
+  assertSourceMapLocationMapping({
+    smc,
+    generatedPosition: [10, 74],
+    authoredPosition: [7, 59],
+    msg: "Identifier 'name' in `[name, '!']` maps to `${name}`",
+  });
+  assertSourceMapLocationMapping({
+    smc,
+    generatedPosition: [10, 10],
+    authoredPosition: [7, 13],
+    msg: "'sayHello' identifier declaration maps to authored declaration",
+  });
+  assertSourceMapLocationMapping({
+    smc,
+    generatedPosition: [9, 37],
```

Comment:
slight concern about maintenance of this should this change. maybe a comment to what the generated code at this position looks like would help?
concern is more so for this one that doesn't have an authored position since the authored code is in git, but the generated code is not.
something like..
```ts
// const lit_template_1 = { h: b_1 `<h1>Hello <?><?></h1>`, parts: [{ type: 2, index: 1 }, { type: 2, index: 2 }] };
//                                     ^
```

Reply:
i guess i also wonder if the .js files should be committed just like the goldens. the sourcemap test really should only be necessary to run when goldens are updated.

Reply:
Agree – will check in the generated js. Should I also check in the `js.map`?